File: leetcode/algorithms/c++/Sudoku Solver/Sudoku Solver.cpp
class Solution
{

private:
    map<vector<int>, vector<vector<int>>> base;
    void init()
    {
        for (int j = 0; j < 9; j++)
        {
            for (int i = 0; i < 9; i++)
            {
                base[{j, 0, 0}].push_back({j, i});
            }
        }

        for (int j = 0; j < 9; j++)
        {
            for (int i = 1; i < 9; i++)
            {
                base[{j, i, 0}] = base[{j, 0, 0}];
            }
        }

        for (int j = 0; j < 9; j++)
        {
            for (int i = 0; i < 9; i++)
            {
                base[{0, j, 1}].push_back({i, j});
            }
        }

        for (int j = 0; j < 9; j++)
        {
            for (int i = 1; i < 9; i++)
            {
                base[{i, j, 1}] = base[{0, j, 1}];
            }
        }

        for (int ii = 0; ii <= 6; ii += 3)
        {
            for (int jj = 0; jj <= 6; jj += 3)
            {
                vector<vector<int>> tmp;
                for (int i = 0; i < 3; i++)
                {
                    for (int j = 0; j < 3; j++)
                    {
                        tmp.push_back({i + ii, j + jj});
                    }
                }
                for (int i = 0; i < 3; i++)
                {
                    for (int j = 0; j < 3; j++)
                    {
                        base[{i + ii, j + jj, 2}] = tmp;
                    }
                }
            }
        }
    }

    vector<char> lega(vector<vector<char>> &t, vector<int> find)
    {
        if (t[find[0]][find[1]] != '.')
            return {};

        vector<int> k(10, 0);

        for (int i = 0; i < 3; i++)
        {
            auto &way = base[{find[0], find[1], i}];
            for (auto &x : way)
            {
                if (t[x[0]][x[1]] != '.')
                    ++k[t[x[0]][x[1]] - '0'];
            }
        }
        vector<char> result;
        for (int i = 1; i <= 9; i++)
        {
            if (k[i] == 0)
                result.push_back(i + '0');
        }
        return result;
    }
    bool oblivius(vector<vector<char>> &t)
    {
        bool isUp = false;
        for (int i = 0; i < 9; i++)
        {
            for (int j = 0; j < 9; j++)
            {
                if (t[i][j] == '.')
                {
                    auto tmp = lega(t, {i, j});
                    if (tmp.size() == 1)
                    {
                        t[i][j] = tmp[0];
                        isUp = true;
                    }
                }
            }
        }
        return isUp;
    }

    bool stop = false;

    bool check(vector<vector<char>> &t)
    {
        for (int i = 0; i < 9; i++)
        {
            for (int j = 0; j < 9; j++)
            {
                if (t[i][j] == '.')
                    return false;
            }
        }
        return true;
    }

    void dfs(vector<vector<char>> &t)
    {
        if (stop)
            return;
        pirnt(t);
        if (check(t))
        {
            stop = true;
            return;
        }

        for (int i = 0; i < 9; i++)
        {
            for (int j = 0; j < 9; j++)
            {
                auto next = lega(t, {i, j});
                if (t[i][j] == '.' && next.size() == 0)
                {
                    //  cout << "ERROR " << i << " " << j << endl;
                    return;
                }
                for (auto &x : next)
                {
                    t[i][j] = x;
                    dfs(t);
                    if (stop)
                        return;
                    //cout <<"BACK " << i << " " << j << " | " << x << endl;
                    t[i][j] = '.';
                }
                if (t[i][j] == '.')
                    return;
            }
        }
    }

    void pirnt(vector<vector<char>> &x)
    { /*
        for(int i = 0; i < 9; i++) {
            for(int j = 0; j < 9; j++) {
                cout << x[i][j];
                if (j % 3 == 2) cout << " ";
            }
            cout << endl;
        }
        cout << endl;*/
    }

public:
    void solveSudoku(vector<vector<char>> &board)
    {
        init();
        //  while(oblivius(board));
        dfs(board);
    }
};

File: leetcode/algorithms/python/Sudoku Solver/Sudoku Solver.py
from typing import List

class Solution:
    base = {}
    stop = False

    def solveSudoku(self, board: List[List[str]]) -> None:
        self.init()
        while (self.oblivius(board)):
            pass
        self.dfs(board)

    def legal(self, t: List[List[str]], find: List[int]) -> List[str]:
        if t[find[0]][find[1]] != ".":
            return []
        k = [0,0,0,0,0,0,0,0,0,0]
        for i in range(3):
            way = self.base[hash(tuple([find[0], find[1], i]))]
            for x in way:
                if t[x[0]][x[1]] != '.':
                    k[int(t[x[0]][x[1]])] += 1
        result = []
        for i in range(1, 10):
            if k[i] == 0:
                result.append(str(i))
        return result
    def check(self, t: List[List[str]]) -> bool:
        for i in range(9):
            for j in range(9):
                if (t[i][j] == '.'):
                    return False
        return True
    def dfs(self, t: List[List[str]]) -> None:
        if (self.stop):
            return
        if (self.check(t)):
            self.stop = True
            return
        for i in range(9):
            for j in range(9):
                next = self.legal(t, [i, j])
                if t[i][j] == '.' and len(next) == 0:
                    return
                for x in next:
                    t[i][j] = x
                    self.dfs(t)
                    if self.stop:
                        return
                    t[i][j] = '.'
                if t[i][j] == '.':
                    return
    def oblivius(self,  t: List[List[str]]) -> bool:
        isUp = False
        for i in range(9):
            for j in range(9):
                if (t[i][j] == '.'):
                    tmp = self.legal(t, [i,j])
                    if (len(tmp) == 1):
                        t[i][j] = tmp[0]
                        isUp = True
        return isUp


    def init(self) -> None:
        for j in range(9):
            for i in range(9):
                key = hash(tuple([j, 0, 0]))
                value = [j, i]
                if key in self.base:
                    self.base[key].append(value)
                else:
                    self.base[key] = [value]
        for j in range(9):
            for i in range(1, 9):
                self.base[hash(tuple([j, i, 0]))] = self.base[hash(tuple([j, 0, 0]))]
        for j in range(9):
            for i in range(9):
                key = hash(tuple([0, j, 1]))
                value = [i, j]
                if key in self.base:
                    self.base[key].append(value)
                else:
                    self.base[key] = [value]
        for j in range(9):
            for i in range(1, 9):
                self.base[hash(tuple([i, j, 1]))] = self.base[hash(tuple([0, j, 1]))]
        for ii in range(0, 7, 3):
            for jj in range(0, 7, 3):
                tmp = []
                for i in range(3):
                    for j in range(3):
                        tmp.append([i + ii, j + jj])
                for i in range(3):
                    for j in range(3):
                        self.base[hash(tuple([i + ii, j + jj, 2]))] = tmp

                
                

    def printBoard(self, board: List[List[str]]) -> None:
        for row in board:
            for cell in row:
                print(cell, end='')
            print("")
        print("")

